# 金融市场情绪专题
PCA

In [1]:
import sys
import os
import cudf
import pandas as pd

# ------------------------------数据集路径----------------------------------#
# /usr/local/miniconda3/envs/Rapids/bin/python
DATASETS_PATH = '/data/DataSets/BW_INDEX_CSMAR/BW_5_CSMAR/'

In [2]:
os.listdir(DATASETS_PATH)

['QX_STOCKRATE[DES][csv].txt',
 'QX_TRM1.csv',
 'QX_FUNDDISCOUNTPREMIUM[DES][csv].txt',
 'QX_DIVIDENDPREMIUM.csv',
 'QX_FUNDDISCOUNTPREMIUM1.csv',
 'QX_TRM.parquet',
 'QX_FUNDDISCOUNTPREMIUM.csv',
 'QX_DIVIDENDPREMIUM[DES][csv].txt',
 'QX_DIVIDENDPREMIUM.parquet',
 'QX_FUNDDISCOUNTPREMIUM.parquet',
 'QX_TRM.csv',
 'QX_TRM[DES][csv].txt',
 'QX_STOCKRATE.csv',
 'QX_IPO[DES][csv].txt',
 'QX_IPO.csv',
 'QX_IPO.parquet']

### 股利溢价
SgnYear [统计年度] - YYYY
Dnum [分红公司数目] - 指当年发生了现金分红的上市公司数量
NDnum [未分红公司数目] - 指当年没有现金分红的上市公司数量
DMB [分红公司平均市值账面比] - 计算公式为：所有分红的公司的市值账面比的平均数
NDMB [未分红公司平均市值账面比] - 计算公式为：所有未分红的公司的市值账面比的平均数
PDND [股利溢价] - 计算公式为：分红公司平均市值账面比-未分红公司平均市值账面比
LogPDND [对数股利溢价] - 计算公式为：股利溢价的对数

In [3]:
def cal_pd():
    df = cudf.read_parquet(f'{DATASETS_PATH}QX_DIVIDENDPREMIUM.parquet')
    return df.rename(columns={'LogPDND': 'lnPD'})[['lnPD']]


df_pd = cal_pd()
df_pd

lnPD
SgnYear        
1991    -0.9005
1992    -0.2720
1993    -0.0592
1994    -0.1019
1995    -0.0237
1996    -0.0648
1997    -0.0639
1998    -0.0479
1999    -0.0426
2000    -0.0986
2001    -0.1214
2002    -0.1240
2003    -0.0583
2004    -0.0422
2005    -0.0260
2006    -0.4844
2007    -0.0814
2008    -0.1156
2009    -1.4346
2010    -0.3186
2011    -0.1881
2012    -0.2554
2013    -0.2100
2014    -0.3537
2015    -0.2930
2016    -0.2139
2017    -0.1705
2018    -0.1669
2019    -0.1055
2020    -0.0270
2021    -0.0318

### 股票融资占股票债券融资比例
SgnYear [统计年度] - YYYY-MM
IPO [A股首次发行金额] - 当年IPO的A股融资额
APublic [A股公开增发金额] -
APrivate [A股非公开增发金额] -
Rs [A股配股金额] -
Refinance [A股再筹资金额] - 通过增发配股等方式融资额
Ashares [A股股票融资总额] - 计算公式为：IPO+再融资之和
Bond [债券市场筹资总额] - 计算公式为：债券市场各类债券发行的总额
SRate [股票融资占股票债券融资比例(%)] - 计算公式为：股票融资*100/(股票融资+债券融资)

In [9]:
# df_r = pd.read_csv(f'{DATASETS_PATH}QX_STOCKRATE.csv')
# df_r

SgnYear      IPO  APublic  APrivate       Rs  Refinance   Ashares  \
0      1990     5.86     0.00      0.00     0.00       0.00      5.86   
1      1991     2.39     2.48      0.00     0.00       2.48      4.87   
2      1992   153.42     0.00      0.00     0.00       0.00    153.42   
3      1993   370.50     0.00      0.00     1.95       1.95    372.45   
4      1994   186.47     7.68      0.00    12.29      19.97    206.44   
5      1995    25.79     0.00      0.00    25.41      25.41     51.20   
6      1996   224.91     0.00      0.00    56.17      56.17    281.08   
7      1997   618.39     0.00      0.00   186.96     186.96    805.35   
8      1998   405.32    50.54      0.00   327.44     377.98    783.30   
9      1999   505.03    55.10      4.65   319.62     379.37    884.40   
10     2000   852.05   143.73      0.00   492.63     636.36   1488.41   
11     2001   614.86   193.05      0.00   430.64     623.69   1238.55   
12     2002   499.48   164.69      0.00    56.61     221.30    720.78   
13     2003   473.01   116.15      0.00    76.52     192.67    665.68   
14     2004   361.05   159.74      0.00   104.77     264.51    625.56   
15     2005    57.63   278.78      0.00     2.62     281.40    339.03   
16     2006  1280.86   111.32    896.57     4.32    1012.21   2293.07   
17     2007  4894.84   616.89   2252.22   227.70    3096.81   7991.65   
18     2008  1063.83   576.38   1618.04   151.57    2345.99   3409.82   
19     2009  1878.98   261.69   2660.40   105.97    3028.06   4907.04   
20     2010  4793.36   377.14   3246.62  1438.22    5061.98   9855.34   
21     2011  2824.43   288.79   4287.24   421.96    4997.99   7822.42   
22     2012  1060.40   105.20   3385.16   121.00    3611.36   4671.76   
23     2013     0.00    80.43   3614.06   475.73    4170.22   4170.22   
24     2014   668.89     3.65   7959.96   137.97    8101.58   8770.47   
25     2015  1576.39     0.00  15088.23    42.34   15130.57  16706.96   
26     2016  1496.08     0.00  18520.64   298.51   18819.15  20315.23   
27     2017  2301.09     0.00  13328.93   162.96   13491.89  15792.98   
28     2018  1378.15     0.00   8918.41   228.32    9146.73  10524.88   
29     2019  2489.81    89.50   9480.04   133.88    9703.42  12193.23   
30     2020  4624.58    25.71   8505.83   512.97    9044.51  13669.09   
31     2021  5323.73     0.00   9334.22   493.35    9827.57  15151.30   

         Bond  SRate  
0       93.49   5.90  
1      199.09   2.39  
2      395.64  27.94  
3      304.54  55.02  
4      418.86  33.01  
5      376.89  11.96  
6     1883.26  12.99  
7      835.22  49.09  
8      104.10  88.27  
9      424.10  67.59  
10     325.50  82.06  
11    1085.00  53.30  
12    1775.50  28.87  
13    2091.57  24.14  
14    3098.13  16.80  
15    5534.80   5.77  
16    7480.08  23.46  
17    6656.68  54.56  
18    8607.95  28.37  
19   18267.66  21.17  
20   20225.83  32.76  
21   16171.49  32.60  
22   21687.94  17.72  
23   22336.97  15.73  
24   26774.97  24.67  
25   71716.38  18.89  
26  115452.63  14.96  
27   95810.06  14.15  
28  116760.42   8.27  
29  133264.86   8.38  
30  205889.03   6.23  
31  202870.10   6.95

### 基金折溢价率CEFD
Symbol [基金代码] -
TradingDate [交易日期] -
CategoryID [基金类别] - S0601=股票型基金;S0602=债券型基金;S0603=货币型基金;S0604=混合型基金;S0605=FOF;S0606=股指期货型基金;S0699=其他
FundTypeID [基金运作方式] - S0501=契约型开放式;S0502=契约型封闭式
IsETF [是否ETF] - 1=是；2=否
IsLOF [是否LOF] - 1=是；2=否
IsIndexFund [是否指数基金] - 1=是；2=否
IsInnovative [是否创新型基金] - 1=是；2=否
NAV [基金份额净值] -
AccumulativeNAV [基金份额累计净值] -
ClosePrice [收盘价] -
CovertRate [溢价率(%)] - 计算公式为：（收盘价-基金份额净值）*101/基金份额净值
TurnoverRate [换手率] -

In [5]:

def cal_cefd():
    """
    计算封闭式基金溢价
    """

    '''
    保留 封闭式基金
    剔除 LOF基金,创新型基金,股票指数基金
    '''

    df = (
            cudf.from_pandas(
                    cudf.read_parquet(f'{DATASETS_PATH}QX_FUNDDISCOUNTPREMIUM.parquet').sort_index().to_pandas()
                    .query(" FundTypeID=='S0502'& IsLOF==2 & IsInnovative==2 & IsIndexFund==2")
            )
    )

    '''
    CovertRate 平均加权
    '''
    df['DayAvgCR'] = df.groupby(level=['TradingDate'])['CovertRate'].transform('mean')

    '''
    时间序列
    '''
    df = df.groupby(level=['TradingDate']).first()

    return df.rename(columns={'DayAvgCR': 'CEFD'})[['CEFD']]


df_cefd = cal_cefd()
df_cefd

CEFD
TradingDate            
19980430     101.500000
19980630      60.272500
19980731      41.500000
19980831      28.418000
19980930      31.054000
...                 ...
20221216       4.303333
20221219       1.501667
20221220       0.239091
20221221       0.583000
20221222       0.293333

[2637 rows x 1 columns]

### 市场换手率
TradingDate [交易日期] -
MarketType [市场类型] - 1=上证A股市场；2=上证B股市场；4=深证A股市场；8=深证B股市场；6=主板；5=沪深A股市场；7=中小板；10=沪深B股市场；15=沪深AB股市场；16=创业板；20=深证A股和创业板；21=沪深A股和创业板；31=沪深AB股和创业板；32=科创板；33=上证A股和科创板；37=沪深A股和科创板；47=沪深AB股和科创板；53=沪深A股和创业板和科创板；63=沪深AB股和创业板和科创板；64=北证A股市场；69=沪深京A股市场；79=沪深京AB股市场；85=沪深京A股和创业板；95=沪深京AB股和创业板；101=沪深京A股和科创板；111=沪深京AB股和科创板；117=沪深京A股和创业板和科创板；127=沪深京AB股和创业板和科创板；
Parameter [参数值] - 字段说明见说明书“附录 参数说明”
Unit [参数单位] -
TurnoverRate1 [换手率(总股本)] - 计算公式为：交易量/总股本
TurnoverRate2 [换手率(流通股本)] - 计算公式为：交易量/流通股本

In [6]:
def transform():
    """
    转换函数
    """
    df_turn = pd.concat([pd.read_csv(f'{DATASETS_PATH}QX_TRM.csv'), pd.read_csv(f'{DATASETS_PATH}QX_TRM1.csv')])
    df_turn['TradingDate'] = df_turn['TradingDate'].str.replace('-', '').astype('uint32')
    df_turn = df_turn.set_index(['TradingDate', 'MarketType']).sort_index()
    df_turn.to_parquet(f'{DATASETS_PATH}QX_TRM.parquet')
    return df_turn


def cal_turn():
    """
    计算换手率
    :return:
    """
    df = cudf.read_parquet(f'{DATASETS_PATH}QX_TRM.parquet').query("MarketType==5")
    return df.reset_index(['MarketType']).rename(columns={'TurnoverRate2': 'TURN'})[['TURN']]


cal_turn()

TURN
TradingDate         
19901219     0.00062
19901220     0.00012
19901221     0.00003
19901224     0.00004
19901225     0.00001
...              ...
20221221     0.08464
20221221     0.18570
20221221     0.52087
20221221     1.74546
20221221     2.35661

[54588 rows x 1 columns]

### IPO股票信息
Symbol [股票代码] -
ListedDate [上市日期] -
ListedYear [上市年份] -
ABSymbol [AB股交叉码] - 指公司同时发行了A和B股的情况
HSymbol [H股交叉码] - 指的是公司同时发行了H股的情况
EstablishDate [公司成立日期] -
IpoDate [首次招股日期] -
CompanyListedDate [公司上市日期] -
IsIPO [是否首次发行] - A=是；B=否
TotalShares [实际发行总量] -
IssuePrice [发行价格] -
CurrencyCode [发行价格币种] -
ClosePrice [上市首日收盘价] -
ReturnRate [上市首日回报率] -
TurnoverRate1 [上市首日换手率(总股本)] -
TurnoverRate2 [上市首日换手率(流通股本)] -
PE [上市首日市盈率] -
PB [上市首日市净率] -

In [7]:
def cal_ipo():
    """
    计算IPO信息
    """
    df = cudf.read_parquet(f'{DATASETS_PATH}QX_IPO.parquet')

    '''
    计算平均首日收益
    '''
    df['DayAvgRR'] = df.groupby(level=['ListedDate'])['ReturnRate'].transform('mean')
    df['DayTotalN'] = df.groupby(level=['ListedDate'])['ReturnRate'].transform('count')
    df = df.groupby(level=['ListedDate']).first()

    return df.rename(columns={'DayAvgRR': 'RIPO', 'DayTotalN': 'NIPO'})[['RIPO', 'NIPO']]


df_ipo = cal_ipo()
df_ipo

RIPO  NIPO
ListedDate                   
19901210          0.803     1
19901219    48.89698229     7
19910114          14.98     1
19910129          13.58     1
19910403          0.225     1
...                 ...   ...
20221216      0.1284755     2
20221221       0.156567     1
20221222       0.157636     2
20221223           <NA>     0
20221226           <NA>     0

[2615 rows x 2 columns]

## 合成

In [32]:
# df_pca = df_cefd.to_pandas().join(df_ipo.to_pandas())
pd.merge(df_cefd.to_pandas().reset_index(), df_ipo.to_pandas().reset_index(), left_on='TradingDate', right_on='ListedDate', how='left')
df_pca

CEFD      RIPO  NIPO
TradingDate                            
19980430     101.500000       NaN   NaN
19980630      60.272500  1.081585   1.0
19980731      41.500000       NaN   NaN
19980831      28.418000       NaN   NaN
19980930      31.054000       NaN   NaN
...                 ...       ...   ...
20221216       4.303333  0.128475   2.0
20221219       1.501667       NaN   NaN
20221220       0.239091       NaN   NaN
20221221       0.583000  0.156567   1.0
20221222       0.293333  0.157636   2.0

[2637 rows x 3 columns]